packages/console/src/hooks/use-tenants.ts

```diff
@@ -0,0 +1,86 @@
+import { useLogto } from '@logto/react';
+import { type TenantInfo } from '@logto/schemas';
+import { type Optional, trySafe } from '@silverhand/essentials';
+import { useCallback, useContext, useEffect, useMemo, useState } from 'react';
+import useSWR from 'swr';
+
+import { useCloudApi } from '@/cloud/hooks/use-cloud-api';
+import { TenantsContext } from '@/contexts/TenantsProvider';
+
+import useSwrFetcher from './use-swr-fetcher';
+
+const useTenants = () => {
+  const cloudApi = useCloudApi();
+  const { signIn, getAccessToken } = useLogto();
+  const {
+    tenants,
+    setTenants: originalSetTenants,
+    currentTenantId,
+    setCurrentTenantId,
+    setIsSettle,
+    navigate,
+  } = useContext(TenantsContext);
+  const [error, setError] = useState<Error>();
+
+  const fetcher = useSwrFetcher<TenantInfo[]>(cloudApi);
+  const {
+    data: availableTenants,
+    error: requestError,
+    mutate,
+  } = useSWR<TenantInfo[], Error>('/api/tenants', fetcher);
+  const isLoading = Boolean(!availableTenants && !requestError);
+  const isLoaded = Boolean(availableTenants && !requestError);
+
+  const setTenants = useCallback(
+    (tenants: TenantInfo[]) => {
+      originalSetTenants(tenants);
+      void mutate(tenants);
+    },
+    [mutate, originalSetTenants]
+  );
+
+  useEffect(() => {
+    if (availableTenants) {
+      setTenants(availableTenants);
+    }
+    if (requestError) {
+      setError(requestError);
+    }
+  }, [availableTenants, requestError, setTenants]);
+
+  const validate = useCallback(
+    async (tenant: TenantInfo) => {
+      const { id, indicator } = tenant;
+      if (await trySafe(getAccessToken(indicator))) {
+        setIsSettle(true);
+      } else {
+        void signIn(new URL(`/${id}/callback`, window.location.origin).toString());
+      }
+    },
+    [getAccessToken, setIsSettle, signIn]
+  );
+
+  const currentTenant: Optional<TenantInfo> = useMemo(() => {
+    return tenants?.find(({ id }) => id === currentTenantId);
+  }, [currentTenantId, tenants]);
+
+  useEffect(() => {
+    if (currentTenant) {
+      void validate(currentTenant);
+    }
+  }, [currentTenant, navigate, setCurrentTenantId, tenants, validate]);
+
+  return {
+    api: cloudApi,
+    tenants,
+    setTenants,
+    currentTenantId,
+    currentTenant,
+    isLoaded,
+    isLoading,
+    error,
+    setError,
+  };
```

Comment:
We might not need to expose `setTenants` and `setError`. Also, it seems we can get `currentTenantId` from `currentTenant`, unless you tell me they are not always consistent (sometime you have currentTenantId but the currentTenant is undefined).